Mathlib/Algebra/Homology/EulerCharacteristic.lean

```diff
@@ -0,0 +1,79 @@
+/-
+Copyright (c) 2025 Jesse Alama. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jesse Alama
+-/
+import Mathlib.Algebra.Homology.HomologicalComplex
+import Mathlib.Algebra.Homology.ShortComplex.HomologicalComplex
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.LinearAlgebra.Dimension.Finrank
+import Mathlib.Topology.Algebra.InfiniteSum.Basic
+
+/-!
+# Euler characteristic of chain complexes
+
+This file defines the Euler characteristic of a chain complex.
+
+## Main definitions
+
+* `ChainComplex.boundedEulerChar`: The Euler characteristic of a chain complex up to index `n`,
+  defined as the alternating sum `∑_{k=0}^{n-1} (-1)^k rank(C_k)` of the ranks of the chain groups.
+* `ChainComplex.eulerChar`: The Euler characteristic as an infinite sum over all natural numbers,
+  defined for chain complexes that eventually vanish (i.e., have finrank 0 beyond some point).
+
+## Main results
+
+* `ChainComplex.boundedEulerChar_zero`: The bounded Euler characteristic at 0 is 0 (empty sum).
+* `ChainComplex.boundedEulerChar_succ`: Recursive formula for the bounded Euler characteristic.
+* `ChainComplex.eulerChar_eq_boundedEulerChar`: The Euler characteristic equals the bounded
+  Euler characteristic for sufficiently large bounds.
+
+-/
+
+namespace ChainComplex
+
+variable {R : Type*} [Ring R]
+
+/-- The bounded Euler characteristic of a chain complex up to index `n`, defined as the alternating
+sum `∑_{k=0}^{n-1} (-1)^k rank(C_k)` of the ranks of the chain groups from index 0 to n-1.
+This is most commonly used when the base ring is a field (where rank means dimension)
+or a principal ideal domain. -/
+noncomputable def boundedEulerChar (C : ChainComplex (ModuleCat R) ℕ) (n : ℕ) : ℤ :=
+  ∑ k ∈ Finset.range n, (-1 : ℤ)^k * Module.finrank R (C.X k)
```

Comment:
This may not be something you're interested in doing, but this could be made massively more general by defining the function from ChainComplex C Nat to the grothendieck group (whether split or exact or triangulated!) of C for any C, not just ModuleCat R.